### ISSUE TITLE
Wayland with WayVNC crashes with no proper recovering

### ISSUE THREAD

**Issue by JohnFarl** [raspberrypi/bookworm-feedback#108]
Problem description:
I use Raspberry mostly with VNC and I'm facing frequent crashes of desktop environment that cause VNC disconnection and loss of control.
If I try to power on HDMI display after these crashes, I see randomly one of these situation:
- black screen
- totally  frozen display, not even mouse cursor moves
- a Raspberry login menu, that is weird considering that I use it with autologin enabled and so it should login automatically, in addition such login menu cannot be used from VNC because at this point VNC connection is already unusable till next OS reboot.

Expected result:
- A more stable environment to avoid crashes in first place that are too frequent despite a clean Raspberry Pi OS image setup.
- Better crash handling with automatic recover of desktop environment and VNC control if for some reason a crash occur.

Device: Pi4 8GB

**Comment by any1** [comment]
Might `journalctl` shed some light on what's actually happening?

Although, in case wayvnc is crashing, I'm not sure if the logs would be very useful given that Pi OS is still running an old version. There are some bugs that have been fixed since the previous release.

**Comment by JohnFarl** [comment]
@any1 
These are the anomalies I have in log that could be related to the issue

```
ktop.portal.Documents[39832]: Ignoring invalid max threads value 4294967295 > max (100000)

pipewire[35788]: spa.v4l2: '/dev/video14' VIDIOC_ENUM_FRAMEINTERVALS: Inappropriate ioctl for device
pipewire[35788]: spa.v4l2: '/dev/video15' VIDIOC_ENUM_FRAMEINTERVALS: Inappropriate ioctl for device
pipewire[35788]: spa.v4l2: '/dev/video21' VIDIOC_ENUM_FRAMEINTERVALS: Inappropriate ioctl for device
pipewire[35788]: spa.v4l2: '/dev/video22' VIDIOC_ENUM_FRAMEINTERVALS: Inappropriate ioctl for device
pipewire[35788]: spa.v4l2: '/dev/video14' VIDIOC_ENUM_FRAMEINTERVALS: Inappropriate ioctl for device


wireplumber[38043]: <WpPortalPermissionStorePlugin:0x5588a84d90> Failed to call Lookup: GDBus.Error:org.fre>
wireplumber[38043]: <WpPortalPermissionStorePlugin:0x5588a84d90> Failed to call Lookup: GDBus.Error:org.fre>
wireplumber[38043]: <WpPortalPermissionStorePlugin:0x5588a84d90> Failed to call Lookup: GDBus.Error:org.fre>

kernel: vc4-drm gpu: swiotlb buffer is full (sz: 389120 bytes), total 32768 (slots), used 27 (slots)
kernel: vc4-drm gpu: swiotlb buffer is full (sz: 442368 bytes), total 32768 (slots), used 1 (slots)
kernel: vc4-drm gpu: swiotlb buffer is full (sz: 561152 bytes), total 32768 (slots), used 7 (slots)
kernel: vc4-drm gpu: swiotlb buffer is full (sz: 274432 bytes), total 32768 (slots), used 5 (slots)
kernel: vc4-drm gpu: swiotlb buffer is full (sz: 565248 bytes), total 32768 (slots), used 5 (slots)

kernel: v3d fec00000.v3d: MMU error from client CLE (4) at 0x12a1000, pte invalid

systemd[1063]: xdg-desktop-portal-gtk.service: Failed with result 'exit-code'.
```

**Comment by hrafnkelle** [comment]
A "reliable" way for me to reproduce this VNC crash is to run wsjtx for a while. VNC  will usually crash within a 10 minutes and no way to reconnect to the RPi.

**Comment by any1** [comment]
There's a chance that this problem might be resolved in the newest version. So, if you're tired of waiting for it to be packaged, you could compile it from source. If that doesn't fix the problem, the stdout and stderr output from wayvnc along with a backtrace would be very useful.

**Comment by andrum993** [comment]
Ideally the wayvnc service unit file would configure systemd so it attempted to restart wayvnc if it detects the wayvnc server has crashed, rather than relying on a user going in and manually rstarting it.

**Comment by any1** [comment]
> Ideally the wayvnc service unit file would configure systemd so it attempted to restart wayvnc if it detects the wayvnc server has crashed, rather than relying on a user going in and manually rstarting it.

Yes, this would be a more robust way of running it than xdg-autostart. It could even be started as a systemd user service.

The best way would, of course, be to modify wayvnc so that it can run detached from the compositor and start it as a system service. Then, when a compositor comes along, it would attach to the compositor. I've experimented with this a bit; just shuffling around some code and not taking into account stability or any kind of resource management, I was able to get it working within quite a short time-frame. However, finishing this is pretty low on my list of priorities.

**Comment by andrum993** [comment]
I hadn't looked at how wayvnc was being managed at present - I was just assuming it was getting run as some sort of systemd service - sounds like that's not the case.

**Comment by andrum993** [comment]
And RealVNC will be coming along at some point with their Wayland-compatible VNC server, so that will probably end up going back in as the default VNC server on Raspberry Pi OS.

**Comment by hrafnkelle** [comment]
I hooked up a monitor to my rpi and reproduced the issue. I don't think the issue is in wayvnc. The whole desktop environment goes down and respawns. VNC is not activated again until after I re-log into the desktop (auto login only seems to work on startup?)
